Cache supported tags for wheels.

mauritsvanrees opened the pull request:
## Summary of changes

In `wheel.py:Wheel.is_compatible` we called `sys_tags` each time to get the list of supported tags. Now we call this once, and cache the result in a simple variable. This drastically speeds up checking the compatibility of hundreds of wheels.

Closes #3804

### Pull Request Checklist
- [x] Changes have tests
- [x] News fragment added in [`changelog.d/`].
  _(See [documentation][PR docs] for details)_


[`changelog.d/`]: https://github.com/pypa/setuptools/tree/master/changelog.d
[PR docs]:
https://setuptools.pypa.io/en/latest/development/developer-guide.html#making-a-pull-request

abravalheri commented:
Hi @mauritsvanrees, thank you for the contribution and sorry for delay in reviewing...

I noticed that you used a global var for the implementation and some patching/mocking/stubbing capabilities for the tests.

What do you think about `functools.lru_cache(maxsize=None)`? (I think `functools.cache` would be better, but it is not available before 3.9 - and by looking at the [implementation](https://github.com/python/cpython/blob/3.11/Lib/functools.py#L651-L653) they are equivalent).

It seems also more elegant for testing because you can call `clear_cache` and there seems to be some evidence that it is [faster](https://discuss.python.org/t/reduce-the-overhead-of-functools-lru-cache-for-functions-with-no-parameters/3956/7) because it is [written in C](https://discuss.python.org/t/reduce-the-overhead-of-functools-lru-cache-for-functions-with-no-parameters/3956/13).


---
Small note:

I don't know if `setuptools/wheel.py` is going to be maintained for long after we completely remove `install/easy_install`, but I understand that this is going to be a different discussion, which should involve the Buildout team that heavily uses it.

Since I could not find any references in the docs about it being deprecated, my assumption is that it is OK to add non-disruptive improvements.

mauritsvanrees commented:
Hi @abravalheri. Thanks for the review and sorry for the delay in acting on it.

> What do you think about `functools.lru_cache(maxsize=None)`?

Good point.  I should probably use that function more often.
I have added a commit to do this.  (And rebased on main and force pushed.)

This change has no effect on the timing of my small test script in issue #3804: the significant performance increase remains.

abravalheri commented on setuptools/tests/test_wheel.py:
```diff
@@ -612,8 +612,17 @@ def sys_tags():
-        'onnxruntime-0.1.2-cp36-cp36m-manylinux1_x86_64.whl').is_compatible()
+    # Clear the supported tags cache, otherwise the sys_tags monkeypatch
+    # has no effect.
+    setuptools.wheel._supported_tags.cache_clear()
```
The errors in the CI are probably related to the fact that the function is `_get_supported_tags` instead of `_supported_tags`.

mauritsvanrees replied:
Strange, I saw the tests passing locally with my code.  Well, there were a few failures, but they were there on the main branch as well.  Maybe I missed something in the big test output.
Your changes are good.  Thanks!

mauritsvanrees commented:
Thanks for your support @abravalheri!
